external/urlfrontier/src/main/java/com/digitalpebble/stormcrawler/urlfrontier/Spout.java

```diff
@@ -50,56 +51,67 @@ public class Spout extends AbstractQueryingSpout {
+        maxBucketNum = ConfUtils.getInt(stormConf, URLFRONTIER_MAX_BUCKETS_KEY, 10);
 
         // initialise the delay requestable with the timeout for messages
         delayRequestable =
-                ConfUtils.getInt(conf, Config.TOPOLOGY_MESSAGE_TIMEOUT_SECS, delayRequestable);
+                ConfUtils.getInt(stormConf, Config.TOPOLOGY_MESSAGE_TIMEOUT_SECS, delayRequestable);
 
         // then override with any user specified config
         delayRequestable =
-                ConfUtils.getInt(conf, "urlfrontier.delay.requestable", delayRequestable);
+                ConfUtils.getInt(stormConf, URLFRONTIER_DELAY_REQUESTABLE_KEY, delayRequestable);
 
-        LOG.info("Initialisation of connection to URLFrontier service on {}", address);
+        // host and port of URL Frontier(s)
+        List<String> addresses = ConfUtils.loadListFromConf(URLFRONTIER_ADDRESS_KEY, stormConf);
+
+        // Selected address
+        String address;
+        switch (addresses.size()) {
+            case 0:
+                LOG.debug("{} has no addresses.", URLFRONTIER_ADDRESS_KEY);
+                address = null;
+                break;
+            case 1:
+                LOG.debug(
+                        "{} with a size of {} is used.", URLFRONTIER_ADDRESS_KEY, addresses.size());
+                address = addresses.get(0);
+                break;
+            default:
+                int totalTasks = context.getComponentTasks(context.getThisComponentId()).size();
+                if (totalTasks != addresses.size()) {
+                    String message =
+                            "Not one task per frontier node. "
+                                    + totalTasks
+                                    + " vs "
+                                    + addresses.size();
+                    LOG.error(message);
+                    throw new RuntimeException(message);
+                }
+                int nodeIndex = context.getThisTaskIndex();
+                Collections.sort(addresses);
+                address = addresses.get(nodeIndex);
+        }
 
-        // add the default port if missing
-        if (!address.contains(":")) {
-            address += ":7071";
+        if (address == null) {
+            channel =
+                    ManagedChannelToolkit.createChannel(
+                            ConfUtils.getString(
+                                    stormConf, URLFRONTIER_HOST_KEY, URLFRONTIER_DEFAULT_HOST),
+                            ConfUtils.getInt(
+                                    stormConf, URLFRONTIER_PORT_KEY, URLFRONTIER_DEFAULT_PORT));
+        } else {
+            channel = ManagedChannelToolkit.createChannel(address);
         }
 
-        channel = ManagedChannelBuilder.forTarget(address).usePlaintext().build();
-        frontier = URLFrontierGrpc.newStub(channel);
+        frontier = URLFrontierGrpc.newStub(channel).withWaitForReady();
```

Comment:
what does the waitForReady do?

Reply:
`waitForReady` ensures that the channel is in the state `READY` before sending an RPC. Due to the fact, that we are in an asynchronous inviroment we could have the case, that we try to send an RPC before the connection is properly initialized.

> gRPC implementations MAY provide a per-RPC option to not fail RPCs as a result of the channel being in TRANSIENT_FAILURE state. Instead, the implementation queues the RPCs until the channel is READY. This is known as "wait for ready." The RPCs SHOULD still fail before READY if there are unrelated reasons, such as the channel is SHUTDOWN or the RPC's deadline is reached.

https://github.com/grpc/grpc/blob/master/doc/wait-for-ready.md

Reply:
thanks, I would have expected this to be the case by default but good to have now